### PR TITLE
Avoid macro redefined warning

### PR DESCRIPTION
**Pull request by catap**
We redefines some macros on FreeBSD and OpenBSD, and it trigers some warning which may lead to failure when `-Werror` is used.

### CODE REVIEW

File: posixlib/src/main/resources/scala-native/unistd.c [hunk 1]
```diff
@@ -26,12 +26,40 @@
  * prevent an application developer from getting the parts of an application
  * which do not actually use the constants from running.
  */
+
+#ifdef _XOPEN_VERSION
+#undef _XOPEN_VERSION
+#endif
 #define _XOPEN_VERSION 0
```
Comment (WojciechMazur) [inline posixlib/src/main/resources/scala-native/unistd.c:33]: I guess it would be better to keep original macro
```suggestion
#ifndef _XOPEN_VERSION
#define _XOPEN_VERSION 0
#endif
```

Comment (catap) [reply posixlib/src/main/resources/scala-native/unistd.c:33]: Not sure, because only part of constant are missed and all of them is switched of.

Let me ping @LeeTibbert as well.

**Comment by ekrich** [comment]
I think there was a discussion about this and we concluded that it was getting ugly so this second approach on top was followed. Perhaps a better approach to using `#undef` and all this stuff on top is to create a macro to replace the following where needed:

```c
#ifdef _PC_ALLOC_SIZE_MIN
    return _PC_ALLOC_SIZE_MIN;
#else
    return 0;
#endif
```

You know, when macros get bad make another :-)

**Comment by catap** [comment]
Ok, I'll take care of this in different way.

But I'd like to finish CI for BSD platform before I move.

**Comment by LeeTibbert** [comment]
@catap 
 
re: `Let me ping @LeeTibbert as well.`

I have been offline for awhile and am about to go offline for another longish (weeks) while.
I'll take a look if I can before I go offline again.

My rapid "wet" reading is that section probably long predates me and has to do with
what is and is_not POSIX and what is intended to be allowed in SN posixlib.

I believe, but would have to re-ground myself, that XOPEN is a superset of POSIX and
that the original intent was that only POSIX and possibly XSI were allowed in posixlib.

Compiling with _XOPEN_VERSION macro defined enabled the compilation of  "intended-to-be-excluded"
XOPEN functions, so it was #undef'd.  I think the #undef was because we do not have absolute control
over the compiler flags passed in; a user could  -D_XOPEN_VERSION.

TL;DR - in case events need action before I can do all my evidence collection & tracing.

I think the ugly but cleanest solution is to put explicit OS specific guards around the _XOPEN_VERSION 
handling.  Again, ugly but gives future devos a sporting chance of understanding the logic chain.

That is, have the existing section for linux, MacOS (Windows?) and a new section 
which does the alternate define of this PR for FreeBSD, NetBSD, and known others as they
may apply.

I would end with a "Unsupported OS" #error pragma.

Each OS gets what they are happy with and what has been tested & shown to work.

**Comment by catap** [comment]
On Fri, 19 Apr 2024 17:28:14 +0200,
LeeTibbert wrote:
> 
> I would end with a "Unsupported OS" #error pragma.
> 
> Each OS gets what they are happy with and what has been tested & shown to work. 
> 

Feel free to push into this branch nessesary changes, and we may discuss it.

Right now no rush with it, we have plenty of time to discus the right solution.

-- 
wbr, Kirill
